patchstore.py

```diff
@@ -0,0 +1,116 @@
+            )
+
+    def add(self, filename: str, content: str, origin: str, timestamp=None):
+        """
+        Add (insert or update) a patch to the store.
+        """
+        if not filename:
+            logging.error('Storing a patch requires a filename to identify the patch')
+            return False
+        if mimetypes.guess_type(filename)[0] not in ('text/x-diff', 'text/x-patch'):
+            logging.error('Filename does not match that of a patch file')
+            return False
+        if not content:
+            logging.error('Empty patch contents')
+            return False
+        if not origin:
+            logging.error('Storing a patch requires a valid origin')
+            return False
+        if timestamp is not None:
+            try:
+                timestamp = datetime.fromisoformat(timestamp).isoformat(sep=' ')
+            except ValueError:
+                logging.error(f'Invalid timestamp format: {timestamp} (expects ISO 8601)')
+                return False
+        else:
+            timestamp = datetime.now().isoformat(sep=' ')
+
+        encoded_content = base64.b64encode(content.encode())
+        checksum = self.compute_checksum(encoded_content)
+
+        if not self.exists(filename, origin):
+            with self.db:
+                self.db.execute(
+                    'INSERT INTO patches(filename, content, checksum, origin, timestamp) '
+                    'VALUES(?, ?, ?, ?, ?)',
+                    (filename, encoded_content, checksum, origin, timestamp)
+                )
+            logging.info(f'Stored new patch: {filename} ({origin})')
+        else:
+            with self.db:
+                self.db.execute(
+                    'UPDATE patches SET content = ?, checksum = ?, timestamp = ? '
+                    'WHERE filename = ? AND origin = ?',
+                    (encoded_content, checksum, timestamp, filename, origin)
+                )
+            logging.info(f'Updated existing patch: {filename} ({origin})')
+
+        return True
+
+    def compute_checksum(self, encoded_content: bytes):
+        hash = hashlib.sha256(encoded_content)
+        return f'{hash.name}:{hash.hexdigest()}'
+
+    def exists(self, filename: str, origin: str):
+        """
+        Check whether a patch matching a given filename and a given origin
+        already exists in the store.
+        """
+        with self.db:
+            cursor = self.db.execute(
+                'SELECT * from patches WHERE filename = ? AND origin = ?',
+                (filename, origin)
+            )
+            if cursor.fetchone():
+                return True
+        return False
+
+    def _search_query(self, query, parameters):
+        with self.db:
+            cursor = self.db.execute(query, parameters)
+            for row in cursor.fetchall():
+                yield Patch(*row)
+
+    def search_by_filename(self, filename: str):
```

Comment:
Can you add a short docstring for the search methods? Right now we don't have documentation generated, but if we go for it at some point it will be useful.

Reply:
Done!